File: python/taflow/metrics/tail_ratio.py
"""Exact whole-history return tail-ratio metric."""

from __future__ import annotations

from typing import Any

from .._native.metrics import TailRatio as _Native
from ._input import as_metric_series


class TailRatio:
    """Compute the magnitude ratio of the 95th and 5th return percentiles.

    TAFlow uses linear-interpolated exact quantiles. The external oracle and
    name mapping is Empyrical Reloaded 0.5.12 ``tail_ratio`` for defined
    results. Scalar warm-up is one usable normalized return. An empty state,
    or a zero-magnitude 5th percentile,
    returns ``None`` rather than emitting a non-finite ratio. Inputs may be
    decimal simple returns, log returns, positive equity levels, or
    non-cumulative period P&L with positive initial capital. The first equity
    level establishes a baseline and does not increment metric length.
    ``nan_policy`` is ``"omit"`` or ``"raise"``; infinities are rejected.
    Mutating lifecycle methods are fluent. Rust retains O(n) observations and
    lazily refreshes the sorted cache; a bulk extension sorts only after all
    accepted observations have been appended. Python performs no arithmetic.
    """

    def __init__(self, nan_policy: str = "omit") -> None:
        """Initialize an empty configured metric."""
        self._state = _Native(nan_policy)

    def from_returns(
        self, returns: Any, *, column: str | None = None
    ) -> "TailRatio":
        """Append chronological decimal simple returns and return this metric."""
        self._state.from_returns(as_metric_series(returns, column=column))
        return self

    def from_log_returns(
        self, log_returns: Any, *, column: str | None = None
    ) -> "TailRatio":
        """Append chronological log returns and return this metric."""
        self._state.from_log_returns(as_metric_series(log_returns, column=column))
        return self

    def from_equity(
        self, equity: Any, *, column: str | None = None
    ) -> "TailRatio":
        """Append chronological positive equity levels and return this metric."""
        self._state.from_equity(as_metric_series(equity, column=column))
        return self

    def from_pnl(
        self,
        pnl: Any,
        initial_capital: float,
        *,
        column: str | None = None,
    ) -> "TailRatio":
        """Append period P&L using required positive initial capital."""
        self._state.from_pnl(
            as_metric_series(pnl, column=column), float(initial_capital)
        )
        return self

    def append(self, value: float) -> "TailRatio":
        """Append one value in the selected domain and return this metric."""
        self._state.append(float(value))
        return self

    def extend(self, values: Any) -> "TailRatio":
        """Append a chronological series in the selected domain and return this metric."""
        self._state.extend(as_metric_series(values))
        return self

    @property
    def value(self) -> float | None:
        """Return the current ratio, or ``None`` while undefined."""
        return self._state.value

    def compute(self) -> float | None:
        """Return the current exact scalar without replaying processed input."""
        return self._state.compute()

    def reset(self) -> "TailRatio":
        """Clear retained observations, preserve configuration, and return this metric."""
        self._state.reset()
        return self

    def __len__(self) -> int:
        """Return the number of usable normalized returns retained by Rust."""
        return len(self._state)


__all__ = ["TailRatio"]

File: crates/taflow-metrics/src/metrics/tail_ratio.rs
use crate::{
    primitives::ExactOrderStatistics, MetricError, MetricInputKind, MetricInputState, MetricResult,
    NanPolicy,
};

/// Ratio of the absolute 95th-percentile return to the absolute 5th-percentile return.
#[derive(Debug, Clone)]
pub struct TailRatio {
    input: MetricInputState,
    order_statistics: ExactOrderStatistics,
}

impl TailRatio {
    /// Construct an empty exact tail-ratio state.
    pub fn new(nan_policy: NanPolicy) -> MetricResult<Self> {
        Ok(Self {
            input: MetricInputState::unbound(nan_policy),
            order_statistics: ExactOrderStatistics::new(),
        })
    }

    /// Append one chronological observation and return the current ratio.
    pub fn append(&mut self, value: f64) -> MetricResult<Option<f64>> {
        if let Some(simple_return) = self.input.append(value)? {
            self.order_statistics.append(simple_return);
        }
        Ok(self.value())
    }

    /// Append a chronological slice and refresh exact quantiles only once.
    pub fn extend(&mut self, values: &[f64]) -> MetricResult<Option<f64>> {
        for &value in values {
            if let Some(simple_return) = self.input.append(value)? {
                self.order_statistics.append(simple_return);
            }
        }
        Ok(self.value())
    }

    /// Return the current ratio, or `None` when empty or the lower tail is zero.
    pub fn value(&mut self) -> Option<f64> {
        let lower = self
            .order_statistics
            .quantile(0.05)
            .expect("TailRatio uses a valid fixed lower quantile")?;
        let upper = self
            .order_statistics
            .quantile(0.95)
            .expect("TailRatio uses a valid fixed upper quantile")?;
        let lower_magnitude = lower.abs();
        if lower_magnitude == 0.0 {
            None
        } else {
            Some(upper.abs() / lower_magnitude)
        }
    }

    /// Return the current exact scalar without replaying observations.
    pub fn compute(&mut self) -> Option<f64> {
        self.value()
    }

    /// Restore fresh-state behavior while preserving input configuration.
    pub fn reset(&mut self) {
        self.input.reset();
        self.order_statistics.reset();
    }

    /// Return the number of usable normalized returns retained.
    pub fn len(&self) -> usize {
        self.input.len()
    }

    /// Return whether no usable normalized returns have been retained.
    pub fn is_empty(&self) -> bool {
        self.input.is_empty()
    }
}

crate::impl_return_metric_lifecycle!(TailRatio);
